src/Network/MQTT/Client.hs

```diff
@@ -0,0 +1,190 @@
+
+------------------------------------------------------------------------------------------------
+
+receiverSTM :: ClientState -> STM ()
+receiverSTM ClientState{..} = do
+  p <- readTQueue receiverQueue
+  case p of
+    (CONNACK ConnackPacket{..}) -> do
+      assert (connackReturnCode == Accepted)
+        ("Connection is not accepted by Server. Status code: " ++ show connackReturnCode)
+      writeTVar sessionPresent (Just connackSessionPresent)
+    (PINGRESP _) -> do
+      sessionPresentFlag <- readTVar sessionPresent
+      assert (isJust sessionPresentFlag)
+        "Connack should be received first"
+      writeTVar pingResp (Just True)
+    (PUBACK PubackPacket{..}) -> do
+      qosLevelVal <- readTVar qosLevel
+      assert (qosLevelVal == Just QoS1)
+        ("Received PUBACK. Expected: QoS1, actual: " ++ show qosLevelVal)
+    _            -> return ()
+
+receiver :: ClientState -> IO ()
+receiver state = do
+  atomically $ receiverSTM state
+  receiver state
+
+sender :: ClientState -> Packet -> IO ()
+sender ClientState{..} p = do
+  case p of
+    (PINGREQ _)                 -> atomically $ writeTVar pingResp (Just False)
+    (PUBLISH PublishPacket{..}) -> do
+      sessionPresentFlag <- atomically $ readTVar sessionPresent
+      assert (isJust sessionPresentFlag) "Connack should be received first"
+      atomically $ writeTVar qosLevel (Just $ messageQoS publishMessage)
+    _ -> return ()
+  atomically $ writeTQueue senderQueue p
+
+internalSender :: ClientState -> SocketHandle -> IO ()
+internalSender state socketHandle = do
+  packet <- atomically $ readTQueue (senderQueue state)
+  sendPacket socketHandle packet
+  print $ "SENT: " ++ show packet
+  internalSender state socketHandle
+
+internalReceiver :: ClientState -> SocketHandle -> IO ()
+internalReceiver state socketHandle = do
+  (_, ps) <- parseMany (fst socketHandle) (parse parsePacket) (parse parsePacket)
+  print $ "RECEIVED: " ++ show ps
+  atomically $ internalReceiverSTM state ps
+  internalReceiver state socketHandle
+  where
+    internalReceiverSTM :: ClientState -> [Packet] -> STM ()
+    internalReceiverSTM ClientState{..} ps =
+      if null ps
+        then retry
+        else forM_ ps $ writeTQueue receiverQueue
+
+main :: IO ()
+main = do
+  socketHandle <- conn "test.mosca.io" "1883"
+
+  clientState <- atomically initState
+
+  forkIO $ internalSender clientState socketHandle
+  forkIO $ internalReceiver clientState socketHandle
+
+  forkIO $ receiver clientState
+  forkIO $ do
+    sender clientState connectPacket
+--    sender clientState pingReq
+    threadDelay sec
+    sender clientState publishPacketQoS1
+
+  threadDelay $ 3 * sec
+
+------------------------------------------------------------------------------------------------
+
+sec = 1000000
+
+data ClientState =
+  ClientState
+  { senderQueue    :: TQueue Packet
+  , receiverQueue  :: TQueue Packet
+  , qosLevel       :: TVar (Maybe QoS)
```

Comment:
Well, I see. You shouldn't just store single last QoS server required. Instead, you should use packet identifier to match requests with responses.